SDK as defined by Milestone 5 (#86)
nucypher/nucypher-ts: website/docs/CBD.md

---
slug: cbd
---

# Condition Based Decryption

Condition Based Decryption (CBD) is a programmable access control service, in which decryption rights are based on the verified fulfillment of predefined conditions.

> Access conditions can be EVM-based (e.g. does the designated recipient own this NFT), RPC-driven (e.g. did the designated recipient commence active mining before this block) or time-based (e.g. has a preset period of time elapsed, after which the recipient's requests will be ignored).

Reviewer: ```suggestion
Access conditions can be EVM-based (e.g. does the requester own this NFT?), RPC-driven (e.g. has the requester sent more than N transactions?) or time-based (e.g. has a preset period of time elapsed, after which the recipient's requests will be ignored).
```
The mining example is extremely confusing, because for starters, it'd be using the `eth_mining` RPC endpoint but evaluating it for each Ursula's ETH provider client, not for the requester. I propose using an account's transaction count, which is something that it's not accessible from the EVM but that can be queried with the RPC API. 

Also the notion of "designated recipient" it's a bit restrictive, and in a way is the same angle that PRE has. Instead, we should be talking about "requester", who can be anyone that fulfills the conditions.

These conditions are also composable and can be combined in any logical sequence or decision tree.  

> CBD involves splitting the secret (a decryption key) into multiples _shares_ and distributing those among authorized and collateralized node operators ([Stakers](https://threshold.network/earn/staker) in the Threshold network).

Reviewer: ```suggestion
CBD involves splitting a joint secret (a decryption key) into multiples _shares_ and distributing those among authorized and collateralized node operators ([Stakers](https://threshold.network/earn/staker) in the Threshold network).
```

Also, "joint" seems to be more in line with the new codename...

Author: :joy:

## Build a Cohort
A [Cohort](./cohort) defines the collection of nodes that will provide CBD services.
_Threshold_ and _Shares_ are two parameters used to construct a Cohort.
For example, a `3-of-5` Cohort requires a threshold of 3 out of 5 shares to reconstruct the original plaintext data.

We can define a cohort by:

```js
import { Cohort } from '@nucypher/nucypher-ts'

const config = {
    threshold: 3,
    shares: 5,
    porterUri: 'https://porter-ibex.nucypher.community',
};
const newCohort = await Cohort.create(config);
```

We also provide `porterUri`.
[Porter](./Glossary.md#porter) is a service that helps us interact with nodes.
In this case we are using an `ibex` porter which means we're on a testnet.

## Create a Condition

Condition Based Decryption obviously also needs some [Conditions](./conditions).
In this tutorial we will check that the user owns a specific ERC721 NFT.

```js
import { Conditions } from '@nucypher/nucypher-ts'

const NFTOwnership = new Conditions.ERC721Balance({
    contractAddress: '0xBC4CA0EdA7647A8aB7C2061c2E118A18a936f13D',

>     chain: 'rinkeby',

Reviewer 2: ```suggestion
    chain: 'Rinkeby', // Ibex uses Rinkeby testnet
```
AFAIK we expect upper-cased network names

    parameters: [
      "5954"
    ],
});
```
It is possible to compose multiple Conditions into a [ConditionSet](./condition_set), but for now we will only use one:

```js
import { Conditions, ConditionSet } from '@nucypher/nucypher-ts'

const conditions = new ConditionSet([NFTOwnership])
```

## Build a Strategy

We combine our Cohort, Conditions, and any other extra parameters into a [Strategy](./strategy).

```js
import { Strategy } from '@nucypher/nucypher-ts'

const newStrategy = Strategy.create(
      cohort: newCohort,
      startDate: new Date(),
      endDate: new Date(Date.now() + 1000 * 60 * 60 * 24 * 30), // 30 days
      conditionSet: conditions
);
```

Finally, we can deploy this Strategy to the Threshold Network, and begin using Condition Based Decryption:

:::note

Deploying a Strategy requires writing to the blockchain.

> This means both a funded wallet is required and also a connection to the blockain via a `provider` (eg. MetaMask).

Reviewer: ```suggestion
This means both a funded wallet is required and also a connection to the blockchain via a `provider` (eg. MetaMask).
```

:::

```js
import detectEthereumProvider from '@metamask/detect-provider';

const provider = await detectEthereumProvider();
const newDeployed = await newStrategy.deploy('test', provider);
```

Our encrypter and decrypter objects are available by:

```js
const encrypter = newDeployed.encrypter;
const decrypter = newDeployed.decrypter;

const plaintext = 'this is a secret';
const encryptedMessageKit = encrypter.encryptMessage(plaintext);

const decryptedMessage = await decrypter.retrieveAndDecrypt(
    [encryptedMessageKit]
);
```

> At decryption time, the user will be asked to verify their address by signing a message in MetaMask.

Reviewer: ```suggestion
At decryption time, the requester will be asked to verify their address by signing a message in MetaMask.
```

A minimum number – a _threshold_ – of those operators holding the key shares must be online and actively participate in partial decryptions that can subsequently be combined by the requester to reconstruct the original plaintext data.

## Build a Cohort
A [Cohort](./cohort) defines the collection of nodes that will provide CBD services.
_Threshold_ and _Shares_ are two parameters used to construct a Cohort.
For example, a `3-of-5` Cohort requires a threshold of 3 out of 5 shares to reconstruct the original plaintext data.

We can define a cohort by:

```js
import { Cohort } from '@nucypher/nucypher-ts'

const config = {
    threshold: 3,
    shares: 5,
    porterUri: 'https://porter-ibex.nucypher.community',
};
const newCohort = await Cohort.create(config);
```

We also provide `porterUri`.
[Porter](./Glossary.md#porter) is a service that helps us interact with nodes.
In this case, we are using an `ibex` porter which means we're on a testnet.

## Create a Condition

Condition Based Decryption obviously also needs some [Conditions](./conditions).
In this tutorial, we will check that the user owns a specific ERC721 NFT.

```js
import { Conditions } from '@nucypher/nucypher-ts'

> const NFTOwnership = new Conditions.ERC721Balance({

Reviewer 3: Is this supposed to be `Conditions.ERC721Ownership`, and not `ERC721Balance`? `ERC721.ownerOf` is specific NFT ownership, and `ERC721.balanceOf` is any NFT from the contract.

Additionally `ownerOf` takes the token id ("5954") whereas `balanceOf` takes an address - https://docs.openzeppelin.com/contracts/4.x/api/token/erc721#IERC721.
